app_dart/lib/src/request_handlers/github/webhook_subscription.dart

```diff
@@ -0,0 +1,659 @@
+
+import '../../../protos.dart' as pb;
+import '../../model/github/checks.dart' as cocoon_checks;
+import '../../request_handling/authentication.dart';
+import '../../request_handling/body.dart';
+import '../../request_handling/exceptions.dart';
+import '../../request_handling/subscription_handler.dart';
+import '../../service/branch_service.dart';
+import '../../service/cache_service.dart';
+import '../../service/config.dart';
+import '../../service/datastore.dart';
+import '../../service/github_checks_service.dart';
+import '../../service/logging.dart';
+import '../../service/scheduler.dart';
+
+// Filenames which are not actually tests.
+const List<String> kNotActuallyATest = <String>[
+  'packages/flutter/lib/src/gestures/hit_test.dart',
+];
+
+/// List of repos that require check for labels and tests.
+const Set<String> kNeedsCheckLabelsAndTests = <String>{
+  'flutter/engine',
+  'flutter/flutter',
+  'flutter/packages',
+  'flutter/plugins',
+};
+
+final RegExp kEngineTestRegExp = RegExp(r'(tests?|benchmarks?)\.(dart|java|mm|m|cc)$');
+final List<String> kNeedsTestsLabels = <String>['needs tests'];
+
+/// Subscription for processing GitHub webhooks.
+///
+// TODO(chillers): Change this to an actual subscription.
+/// The PubSub subscription is set up here:
+/// https://cloud.google.com/cloudpubsub/subscription/detail/github-webhooks?project=flutter-dashboard&tab=overview
+///
+/// This endpoint enables Cocoon to recover from outages.
+///
+/// This endpoint takes in a POST request with the GitHub event JSON.
+// TODO(chillers): There's potential now to split this into seprate subscriptions
+// for various activities (such as infra vs releases). This would mitigate
+// breakages across Cocoon.
+@immutable
+class GithubWebhookSubscription extends SubscriptionHandler {
+  /// Creates a subscription for processing GitHub webhooks.
+  const GithubWebhookSubscription({
+    required CacheService cache,
+    required Config config,
+    required this.scheduler,
+    this.githubChecksService,
+    this.datastoreProvider = DatastoreService.defaultProvider,
+    required this.branchService,
+    AuthenticationProvider? authProvider,
+  }) : super(
+          cache: cache,
+          config: config,
+          authProvider: authProvider,
+          topicName: 'github-webhooks',
+        );
+
+  /// Cocoon scheduler to trigger tasks against changes from GitHub.
+  final Scheduler scheduler;
+
+  /// To provide build status updates to GitHub pull requests.
+  final GithubChecksService? githubChecksService;
+
+  ///
+  final DatastoreServiceProvider datastoreProvider;
+  final BranchService branchService;
+
+  @override
+  Future<Body> post() async {
+    if (message.data == null || message.data!.isEmpty) {
+      log.warning('GitHub webhook message was empty. No-oping');
+      return Body.empty;
+    }
+
+    final pb.GithubWebhookMessage webhook = pb.GithubWebhookMessage.fromJson(message.data!);
+
+    log.fine('Processing ${webhook.event}');
```

Comment:
What is this log level? I am not familiar with it and when do we want to use it?

Reply:
See https://pub.dev/packages/logging for the full details, but tl;dr is it's similar to what other loggers would have as debug. For our uses, we log every level so it doesn't matter beyond filtering purposes.